Make the working directory configurable

tkotosz opened the pull request:
Changelog:
- make the working directory configurable
- do not delete the working directory if is is not an auto-generated one, instead keep track of the created files and delete those
- only create the working directory if it does exists (so if a real directory specified then that will be used)
- make properties protected to allow extending the context (e.g. adding new steps to create different kind of files which need to be registered in the "files" list to be deleted at the end)
- fixed a bug in the after scenario hook, where it tries to output the inner behat's output even if the inner behat wasn't started. e.g. because the outer scenario fails earlier than the "i run behat" step

Todo: Add some tests for the file cleanup?

Question: Should we delete the working dir if it is an automatically created tmp one?

Update: probably need to be deleted, or we should use `/tmp/behat-test-runner` as default working dir instead of `/tmp/behat-test-runner{randomhash}` to not end up with many tmp dirs.
@elvetemedve  pushed the change to switch to a fixed working directory... what do you think? What was the reason for using random directory name? :thinking:

tkotosz commented:
scrutinizer error is false positive, it thinks that the method expects `object<Symfony\Co...nt\Filesystem\iterable>` while it expects `\iterable`

elvetemedve commented on src/TestRunnerContext.php:
```diff
@@ -120,12 +126,21 @@ public function afterRunTests(AfterScenarioScope $scope)
         $this->documentRoot = $this->workingDirectory .'/document_root';
-        $this->filesystem->mkdir($this->documentRoot, 0770);
+
+        if (!$this->filesystem->exists($featuresDirectory)) {
```
You are checking the existence of the `$featuresDirectory`, but creating `$this->documentRoot`. That's a bug.
It should be like
```php
if (!$this->filesystem->exists($this->documentRoot)) {
```

tkotosz replied:
oopsy :D

tkotosz replied:
fixed

tkotosz commented:
@elvetemedve I have updated the PR:
- fixed a bug in the after scenario hook, where it tries to output the inner behat's output even if the inner behat wasn't started. e.g. because the outer scenario fails earlier than the "i run behat" step
- changed back the tmp working directory name generating logic to use `tempnam`
- added back the removal of the working directory it is is an auto-generated one

{quote}
I see a potential issue with it. If Behat stops sooner than the AfterScenaro hook is executed (e.g. an exception is thrown), then clearWorkingDirectory() won't be called, leaving junk in the filesystem. Since the list of created files would be lost, we couldn't remove those files later.
I would suggest to attempt to remove the files in TestRunnerContext::__destruct() as well as from AfterSceario. The destructor is called even if the normal script execution stops.
{quote}

unfortunately on ` TestRunnerContext::__destruct()` the `$this->filesystem` is already null, since it gets destructed earlier I guess... on the other hand your example of `e.g. an exception is thrown` is wrong, the `AfterScenario` hook is executed even when an exception is thrown, that's why you can hook into checking if the scenario failed or not. So the only way to not execute the `AfterScenario` is to have a fatal error somewhere in one of the behat steps I guess in which case the folder/files would not be cleaned up. BUT that's already the case in the current version as well, so I think solving that is out of scope of this PR.

tkotosz commented on src/TestRunnerContext.php:
```diff
@@ -60,41 +60,51 @@
      */
     public function __construct(
         $browserCommand = null,
+        $workingDirectory = null,
```
should I move this to be the last parameter? would look weird, but that way if someone already extended this class then it would not fail on `parent::__construct`. what do you think?

elvetemedve replied:
I would prefer to add it as the last constructor argument.

tkotosz replied:
moved

elvetemedve commented on spec/Bex/Behat/Context/TestRunnerContextSpec.php:
```diff
@@ -49,6 +49,16 @@ function it_creates_temporary_working_directory(Filesystem $filesystem)
 
+    function it_removes_temporary_working_directory(Filesystem $filesystem)
+    {
+        $tempDir = sys_get_temp_dir() .'/behat-test-runner';
```
It's surprising that this spec passes as the temporary working directory is created by
`tempnam(sys_get_temp_dir(), 'behat-test-runner') === "/tmp/behat-test-runner<random-characters>" !== "/tmp/behat-test-runner"`.

tkotosz replied:
not surprising because `$filesystem->remove(Argument::containingString($tempDir))->shouldBeCalled();` - see `Argument::containingString` - you wrote it! :D

elvetemedve replied:
It's too early morning to me. Sorry.
